File: sem3/map/ToyLanguageInterpreter/src/model/expressions/ArithExpr.java
package model.expressions;

import utils.Heap;
import utils.SymbolTable;
import utils.exceptions.InterpreterException;

/**
 * Created by mirko on 12/10/2016.
 */
public class ArithExpr implements Expression {
    private char operator;
    private Expression first;
    private Expression second;

    public ArithExpr(Expression first, Expression second, char operator) {
        this.operator = operator;
        this.first = first;
        this.second = second;
    }

    @Override
    public int evaluate(SymbolTable<String, Integer> symTable, Heap<Integer, Integer> heap) throws InterpreterException {
        int firstResult = first.evaluate(symTable, heap);
        int secondResult = second.evaluate(symTable, heap);

        switch (operator) {
            case '+':
                return firstResult + secondResult;
            case '-':
                return firstResult - secondResult;
            case '*':
                return firstResult * secondResult;
            case '/': {
                if (secondResult == 0) throw new InterpreterException("error: division by zero");
                return firstResult / secondResult;
            }
            default:
                throw new InterpreterException("error: invalid operator");
        }
    }

    @Override
    public String toString() {
        return first.toString() + operator + second.toString();
    }
}

File: sem3/map/ToySharpLanguageInterpreter/ToySharpLanguageInterpreter/src/model/expressions/ArithExpr.cs
using System;

namespace ToySharpLanguageInterpreter
{
	public class ArithExpr : Expression
	{
		private char op;
		private Expression first;
		private Expression second;

		public ArithExpr(Expression first, Expression second, char op)
		{
			this.op = op;
			this.first = first;
			this.second = second;
		}

		public int evaluate(SymbolTable<string, int> symTable)
		{
			int firstResult = first.evaluate(symTable);
			int secondResult = second.evaluate(symTable);

			switch (op)
			{
				case '+':
					return firstResult + secondResult;
				case '-':
					return firstResult - secondResult;
				case '*':
					return firstResult * secondResult;
				case '/':
					{
						if (secondResult == 0) throw (new InterpreterException("error: division by zero"));
						return firstResult / secondResult;
					}
				default:
					throw (new InterpreterException("error: invalid op"));
			}
		}

		public override string ToString()
		{
			return first.ToString() + op + second.ToString();
		}
	}
}